ui/app/components/secure-variable-form.js

```diff
@@ -110,13 +152,75 @@ export default class SecureVariableFormComponent extends Component {
     }
   }
 
+  //#region JSON Editing
+
+  view = this.args.view;
+  // Prevent duplicate onUpdate events when @view is set to its already-existing value,
+  // which happens because parent's queryParams and toggle button both resolve independently.
+  @action onViewChange([view]) {
+    if (view !== this.view) {
+      set(this, 'view', view);
+      this.translateAndValidateItems(view);
+    }
+  }
+
+  @action
+  translateAndValidateItems(view) {
+    // TODO: move the translation functions in serializers/variable.js to generic importable functions.
+    if (view === 'json') {
+      // Translate table to JSON
+      set(
+        this,
+        'JSONItems',
+        this.keyValues
+          .filter((item) => item.key.trim() && item.value) // remove empty items when translating to JSON
+          .reduce((acc, { key, value }) => {
+            acc[key] = value;
+            return acc;
+          }, {})
+      );
+    } else if (view === 'table') {
+      // Translate JSON to table
+      set(
+        this,
+        'keyValues',
+        A(
+          Object.entries(this.JSONItems).map(([key, value]) => {
+            return {
+              key,
+              value,
+              warnings: EmberObject.create(),
+            };
+          })
+        )
+      );
+    }
+
+    // Reset any error state, since the errorring json will not persist
+    set(this, 'JSONError', null);
+  }
+
+  get stringifiedItems() {
+    return JSON.stringify(this.args.model.items, null, 2);
```

Comment:
`question`:  Didn't you write a helper for this? We could move this to the template if that's the case.